pkg/network/protocols/http/shared_libraries.go

```diff
@@ -35,91 +37,189 @@ func (l *libPath) Bytes() []byte {
+	m     sync.Mutex
+	byID  map[pathIdentifier]*soRegistration
+	byPID map[uint32]*soRegistration
+
+	// if we can't register a uprobe we don't try more than once
+	blacklistByID map[pathIdentifier]struct{}
 }
 
-func newSOWatcher(procRoot string, perfHandler *ddebpf.PerfHandler, rules ...soRule) *soWatcher {
+func newSOWatcher(perfHandler *ddebpf.PerfHandler, rules ...soRule) *soWatcher {
 	allFilters := make([]string, len(rules))
 	for i, r := range rules {
 		allFilters[i] = r.re.String()
 	}
 
 	all := regexp.MustCompile(fmt.Sprintf("(%s)", strings.Join(allFilters, "|")))
 	return &soWatcher{
-		procRoot:   procRoot,
-		hostMount:  os.Getenv("HOST_ROOT"),
-		all:        all,
-		rules:      rules,
-		loadEvents: perfHandler,
+		procRoot:       util.GetProcRoot(),
+		all:            all,
+		rules:          rules,
+		loadEvents:     perfHandler,
+		processMonitor: monitor.GetProcessMonitor(),
 		registry: &soRegistry{
-			procRoot: procRoot,
-			byPath:   make(map[string]*soRegistration),
-			byInode:  make(map[uint64]*soRegistration),
+			byID:          make(map[pathIdentifier]*soRegistration),
+			byPID:         make(map[uint32]*soRegistration),
+			blacklistByID: make(map[pathIdentifier]struct{}),
 		},
 	}
 }
 
+type soRegistration struct {
+	pathID pathIdentifier
+	//mTime  uint64 TODO(nplanel) reload uprobes if the file changed (dnf upgrade)
+
+	refcount     int
+	unregisterCB func(pathIdentifier) error
+}
+
+func (r *soRegistration) Unregister() (cleanup bool) {
+	r.refcount--
+	if r.refcount > 0 {
+		return false
+	}
+	if r.unregisterCB != nil {
+		if err := r.unregisterCB(r.pathID); err != nil {
+			log.Debugf("unregisterCB %s : %s", r.pathID.String(), err)
+		}
+	}
+	return true
+}
+
+func newRegistration(pathID pathIdentifier, unregister func(pathIdentifier) error) *soRegistration {
+	return &soRegistration{
+		pathID:       pathID,
+		unregisterCB: unregister,
+		refcount:     1,
+	}
+}
+
+func (w *soWatcher) processExit(pid uint32) {
+	w.registry.Unregister(pid)
+}
+
 // Start consuming shared-library events
 func (w *soWatcher) Start() {
-	seen := make(map[seenKey]struct{})
-	w.sync()
+	thisPID, _ := util.GetRootNSPID()
+
+	_ = util.WithAllProcs(w.procRoot, func(pid int) error {
+		if pid == thisPID { // don't uprobes ourself
+			return nil
+		}
+
+		// report silently parsing /proc error as this could happen
+		// just exit processes
+		proc, err := process.NewProcess(int32(pid))
+		if err != nil {
+			log.Debugf("process %d parsing failed %s", pid, err)
+			return nil
+		}
+		mmaps, err := proc.MemoryMaps(true)
+		if err != nil {
+			log.Debugf("process %d maps parsing failed %s", pid, err)
+			return nil
+		}
+
+		root := fmt.Sprintf("%s/%d/root", w.procRoot, pid)
+		for _, m := range *mmaps {
+			for _, r := range w.rules {
+				if !r.re.MatchString(m.Path) {
+					continue
+				}
+
+				w.registry.Register(root, m.Path, uint32(pid), r)
+			}
+		}
+
+		return nil
+	})
+
+	err := w.processMonitor.Initialize()
```

Comment:
calling `WithAllProcs` (sync) and only then `w.processMonitor.Initialize()` means that we can loose processes spawned during that time